external-crates/move/crates/move-compiler/src/typing/deprecations.rs

```diff
@@ -0,0 +1,670 @@
+// Copyright (c) The Diem Core Contributors
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    diag,
+    diagnostics::codes::TypeSafety,
+    expansion::ast::{self as E, ModuleIdent},
+    naming::ast::{self as N, StructFields, Type, Type_},
+    parser::ast::{ConstantName, DatatypeName, FunctionName},
+    shared::{
+        known_attributes::{AttributePosition, KnownAttribute},
+        CompilationEnv, Identifier, Name,
+    },
+    typing::{ast as T, visitor::TypingVisitorContext},
+};
+use move_ir_types::location::*;
+use move_proc_macros::growing_stack;
+use std::collections::{BTreeMap, BTreeSet};
+
+const NOTE_STR: &str = "note";
+
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+struct Deprecation {
+    // The source location of the deprecation attribute
+    source_location: Loc,
+    // The type of the member that is deprecated (function, constant, etc.)
+    location: AttributePosition,
+    // The module that the deprecated member belongs to. This is used in part to make sure we don't
+    // register deprecation warnings for members within a given deprecated module calling within
+    // that module.
+    module_ident: ModuleIdent,
+    // Information about the deprecation information depending on the deprecation attribute.
+    deprecation_info: DeprecationInfo,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+enum DeprecationInfo {
+    // #[deprecated]
+    Deprecated,
+    // #[deprecated(note = b"message")]
+    DeprecatedWithNote(String),
+}
+
+struct Deprecations<'env> {
+    env: &'env mut CompilationEnv,
+    deprecated_modules: BTreeMap<ModuleIdent, Deprecation>,
+    deprecated_constants: BTreeMap<(ModuleIdent, ConstantName), Deprecation>,
+    deprecated_functions: BTreeMap<(ModuleIdent, FunctionName), Deprecation>,
+    deprecated_types: BTreeMap<(ModuleIdent, DatatypeName), Deprecation>,
+    // We store the locations of the deprecation warnings bucketed by "named context" (function,
+    // struct, constant), to then merge them into a single warning at the end. This prevents us
+    // from exploding with too many errors within a given named context.
+    deprecation_warnings: BTreeMap<((ModuleIdent, Name), Deprecation), BTreeSet<Loc>>,
+
+    // Information mutated during the visitor to set the current named context we are within in the
+    // visitor.
+    current_mident: Option<ModuleIdent>,
+    current_named_context: Option<Name>,
+}
+
+impl<'env> Deprecations<'env> {
+    // Index the modules and build up the set of members that are deprecated for the program.
+    fn new(env: &'env mut CompilationEnv, prog: &T::Program) -> Self {
```

Comment:
Consider `impl TypingVisitorConstructor for Deprecations`

Reply:
Will give it a shot. For some reason I remember clippy not being happy about this (and it used to be fine with this until the most recent Rust release I believe?), but I might be misremembering.

Reply:
Yea, this is an error in Rust -- it requires you to include the lifetime params now (even if they're just `'_`)